### PR TITLE
Improve performance and correctness of Tiling Patterns

### LINKED ISSUE #1434: Lines become pixelated and not drawn properly
Some of the lines in this PDF are not drawn properly. They appeared pixelated to the background.
http://carto.metro.free.fr/documents/CartoMetroParis.v3.4.pdf

### LINKED ISSUE #2825: Rendering slow, fuzzy and incomplete
In FF 19, this pdf works not.
http://beta.arcig.cz/~bernard/vyuka-m/6krychli.pdf

### LINKED ISSUE #5698: Image element display problem in PDF.js
Dear All,

PDF.js doesn't display an image on my pdf file, this element is in header of my file. I couldn't find the same problem that i have.

my sample will be found in [here](http://www.asidesigner.com/R.pdf) 

Regards,
Sambose

### LINKED ISSUE #6496: Blurry PDF tiling patterns with huge xstep/ystep
Most of the text is blurry in [GNK Studie in Cijfers 2015.pdf](https://github.com/mozilla/pdf.js/files/6870/GNK.Studie.in.Cijfers.2015.pdf) (source: http://www.uu.nl/bachelors/sites/default/files/editor/gnk/documenten/GNK%20Studie%20in%20Cijfers%202015.pdf).

The text is drawn using a tiling pattern ([PDF spec, 8.7.3 Tiling patterns](https://www.adobe.com/content/dam/Adobe/en/devnet/acrobat/pdfs/PDF32000_2008.pdfPDF8.7.3)) with cells of size 99999x99999... PDF.js [shrinks that to 3000x3000](https://github.com/mozilla/pdf.js/blob/82f73dfed51fd39df7b9df5cd422173d30ad427d/src/display/pattern_helper.js#L339-L347), which results in the blurred text.

```
14 0 obj
<< /Type /Pattern /PatternType 1 /PaintType 1 /TilingType 2
/Resources 13 0 R
/BBox [0 0 748.734 682.715]
/XStep 99999
/YStep 99999
/Matrix [1 0 0 -1 0.000 688.819]
```

Expected (viewed using PDFium):
pdfium
PDF.js:
max-pattern-size-3k
If I set [`MAX_PATTERN_SIZE = 16000;`](https://github.com/mozilla/pdf.js/blob/82f73dfed51fd39df7b9df5cd422173d30ad427d/src/display/pattern_helper.js#L291), then it's less blurry but still illegible:
max-pattern-size-16k

If I set `MAX_PATTERN_SIZE` to 99999 (or even lower 20000), then the gradient is not painted at all... Obviously we cannot create a full canvas of 100k x 100k, but we should still be able to draw this somehow.

### LINKED ISSUE #6737: Title of the document is not visible with proper color
Hi All,

we are using PDF.js and observed that following types of documents are not rendering correctly in IE11, Chrome and FF. it lost its color and header tittle.
[Florida Industrial MarketView Q3 2015.pdf](https://github.com/mozilla/pdf.js/files/55597/Florida.Industrial.MarketView.Q3.2015.pdf)

I have open the document from the link http://mozilla.github.io/pdf.js/web/viewer.html and it gives me the same result. find the attached screen.

image

is there any solution for this issue  ?

Regards,
Sanvid

### LINKED ISSUE #7731: PDF image displayed blurred on Firefox
Link to PDF file (or attach file here):
http://stud.usv.ro/~apahomi/paco.pdf

Configuration:
- Web browser and its version: Firefox 47.0.1
- Operating system and its version: Windows 10 Pro 64 bits
- PDF.js version:
- Is an extension: No, it's firefox default build in pdf viewer

Steps to reproduce the problem:
1.  Access the provided link

What is the expected behavior? (add screenshot)
Open the link in chrome and see the pdf or change the default pdf viewer to another pdf app in firefox or simply just save the pdf(even the one with image blured)

What went wrong? (add screenshot)
The main image is displayed blurred on Firefox but is displayed as it should be on other browsers. Even in firefox if I choose to save the pdf displayed blurred and open it later from my local machine with other app, the pdf looks as it should be.

### LINKED ISSUE #7946: pdf.js showing wrong portion of images
Link to PDF file (or attach file here): https://cloud.box.com/s/3idigur8fepcqxto007ykhjca3w0x68i (after following step 1 below)

Configuration:
- Web browser and its version: Chrome 55
- Operating system and its version: Mac OSX El Capitan
- PDF.js version: 1.6.334
- Is an extension: No

Steps to reproduce the problem:
1. Turn on pdf.js viewer by visiting https://cloud.box.com/on/beta
2. Visit https://cloud.box.com/s/3idigur8fepcqxto007ykhjca3w0x68i - the page should look like https://www.screencast.com/t/lYCWRzAsix
3. Go to page 6
4. Note image in bottom right is cropped/positioned at the wrong place compared to when file is downloaded and viewed through Chrome or Adobe Reader, which show the landscape and not just the bottom portion of the chair

What is the expected behavior? (add screenshot)
page 6 - http://www.screencast.com/t/kfdy2ftjIF
page 15 - http://www.screencast.com/t/nk6k6fLz4XI

What went wrong? (add screenshot)
page 6 - http://www.screencast.com/t/N9Pd9Hgha
page 15 - http://www.screencast.com/t/JndDKRYO8

Link to a viewer (if hosted on a site other than mozilla.github.io/pdf.js or as Firefox/Chrome extension): https://cloud.box.com/s/3idigur8fepcqxto007ykhjca3w0x68i (after following step 1 above)

### LINKED ISSUE #8083: Unrecognisable Blurry Image
Link to PDF file: [blurry-background-image.pdf](https://github.com/mozilla/pdf.js/files/786348/blurry-background-image.pdf)

Configuration:
- Web browser and its version: Firefox 51.0.1 (32-bit)
- Operating system and its version: Windows 7
- PDF.js version: Using version bundled with Firefox

Steps to reproduce the problem:
1. View Attached PDF with PDF.js in Firefox

What is the expected behavior? (add screenshot): Adobe Reader correctly shows the image and isn't blurry. PDF.js does not. See screenshot for comparison.

pdfjs_adobe_reader

### LINKED ISSUE #8264: Images not showing in PDF
Link to PDF file (or attach file here):

[FVD1.pdf](https://github.com/mozilla/pdf.js/files/910730/FVD1.pdf)

On all browsers that we've tried (e.g. Chrome for Mac latest) the three images towards the top right of the first page do not appear.

I have tried with the standard viewer:
https://mozilla.github.io/pdf.js/web/viewer.html

What is the expected behavior? (add screenshot)

screen shot 2017-04-10 at 17 21 41

What went wrong? (add screenshot)

screen shot 2017-04-10 at 17 21 30

### LINKED ISSUE #8373: Missing dots in certain PDF file
Link to PDF file (or attach file here):
[Nat Struct Mol Biol 2008 Zenklusen.pdf](https://github.com/mozilla/pdf.js/files/975293/Nat.Struct.Mol.Biol.2008.Zenklusen.pdf)

Configuration:
- Web browser and its version: Chrome 58.0.3029.96
- Operating system and its version: Windows 15063.250
- PDF.js version: 1.7.225
- Is an extension: No

Steps to reproduce the problem:
Have a look at the top right pic on page 6

What is the expected behavior? (add screenshot)
image

What went wrong? (add screenshot)
image

### LINKED ISSUE #8579: Incorrect line position
Link to PDF file (or attach file here):
[document.pdf](https://github.com/mozilla/pdf.js/files/1108578/document.pdf)

Configuration:
- Web browser and its version:  Google Chrome 59.0.3071.109
- Operating system and its version: Windows 7
- PDF.js version: demo (1.8.497)
- Is an extension:

Steps to reproduce the problem:
1. Open document.pdf in pdf.js.

What is the expected behavior? (add screenshot)
The line is under the text. Just like in Abode PDF Reader.
adobe

What went wrong? (add screenshot)
The line is above the text.
pdfjs

Link to a viewer (if hosted on a site other than mozilla.github.io/pdf.js or as Firefox/Chrome extension):

### LINKED ISSUE #9665: Background images not displayed in page margin boxes
Attach (recommended) or Link to PDF file here: https://assets.menutech.com/food_menu.pdf

**Configuration:**
- Web browser and its version: Firefox Quantum 59.0.2 (64-Bit)
- Operating system and its version: Mac OS X 10.13.3
- PDF.js version: browser in-built
- Is a browser extension: no

**Steps to reproduce the problem:**
1. Load example PDF: https://assets.menutech.com/food_menu.pdf
2. Page margin background image is not displayed

**What is the expected behavior?**
Downloading the PDF and opening it with Adobe Reader correctly shows the background image in the @top-center page margin box. -> PDF.js should also display background images in the page margin boxes.

**Other information:**
Weasyprint was used to generate the PDF (github.com/Kozea/WeasyPrint). The CSS code to add the background image is as follows:

```
@page {
    @top-center {
        content: '';
        display: block;
        width: 100%;
        height: 30mm;
        background-image: <LOGO URL>;
        background-repeat: no-repeat;
        background-size: contain;
        background-position: 50% 50%;
        border-bottom: #fff solid 10mm;
    }
}
```

### LINKED ISSUE #9763: Background portrait in PDF not rendered in pdf.js
Attach PDF file here: [tatertots-portrait.pdf](https://github.com/mozilla/pdf.js/files/2060602/tatertots-portrait.pdf)

```
Title:          PowerPoint Presentation
Creator:        Acrobat PDFMaker 18 for PowerPoint
Producer:       Adobe PDF Library 18.9.66
CreationDate:   Sun Jan 21 20:41:25 2018 PST
ModDate:        Thu May 31 17:42:32 2018 PDT
Tagged:         yes
UserProperties: no
Suspects:       no
Form:           AcroForm
JavaScript:     no
Pages:          1
Encrypted:      no
Page size:      1920 x 1080 pts
Page rot:       0
File size:      995873 bytes
Optimized:      no
PDF version:    1.6
```

Configuration:
- Web browser and its version: Chrome OSX 66.0.3359.181, Firefox 60.0.1 (64-bit)
- Operating system and its version: macOS High Sierra 10.13.4
- PDF.js version: Latest in https://mozilla.github.io/pdf.js/web/viewer.html as of 2018-05-31 
- Is a browser extension: no

Steps to reproduce the problem:
1. Open attached file in Mac PDF viewer or via Chrome 66.0.3359.181
2. View attached file in Firefox 60.0.1 or on Chrome 66.0.3359.181 via  https://mozilla.github.io/pdf.js/web/viewer.html

Expected:
tatertots-portrait_good

Actual:
tatertots-portrait_bad

### LINKED ISSUE #9770: pdf background doesn't load correctly
Attach (recommended) or Link to PDF file here:
[sample.pdf](https://github.com/mozilla/pdf.js/files/2062727/sample.pdf)

Configuration:
- Web browser and its version: This issue is on latest versions of IE, firefox & chrome
- Operating system and its version: Windows, macos Sierra 10.13.4
- PDF.js version: Tested on latest version on https://mozilla.github.io/pdf.js/web/viewer.html
- Is a browser extension: No

Steps to reproduce the problem:
1. Load the sample.pdf in acrobat reader or directly in chrome browser.
2. The pdf will show a black  background to the graph.
3. Open the same sample.pdf in  https://mozilla.github.io/pdf.js/web/viewer.html, the black background will not load completly. It will load in a curved shape.

What is the expected behavior? (add screenshot)
The black backgroung should load on full page as in acrobat reader.
What went wrong? (add screenshot)
image

Expected behaviour: in acrobat reader
image

Link to a viewer (if hosted on a site other than mozilla.github.io/pdf.js or as Firefox/Chrome extension):

### LINKED ISSUE #9922: PDF pages get cut off
Attach (recommended) or Link to PDF file here:
https://drive.google.com/file/d/1LC_yAkFXtVza2aWisMNQbyetjv8Fzij-/view?usp=sharing

Configuration:
- Web browser and its version: **is actual for all browsers**
- Operating system and its version:
- PDF.js version: **latest**
- Is a browser extension:

Steps to reproduce the problem:
1. open pdf file by link
2. download file and open on your demo page
3. compare the images

What is the expected behavior? (add screenshot)

What went wrong? (add screenshot) **The right side of the pictures is getting cut off, and the red part at the bottom goes up higher than it should.**

Link to a viewer (if hosted on a site other than mozilla.github.io/pdf.js or as Firefox/Chrome extension):

### LINKED ISSUE #9927: Incorrect pattern positioning
Attach (recommended) or Link to PDF file here:
[pop-bugzilla41462.pdf](https://github.com/mozilla/pdf.js/files/2232812/pop-bugzilla41462.pdf)

poppler issue:
https://bugs.freedesktop.org/show_bug.cgi?id=41462

Configuration:
- Web browser and its version: FF 61.0.1 64bit
- Operating system and its version: W7
- PDF.js version: latest

Steps to reproduce the problem:
1. start pdf.js viewer url
2. open file

What is the expected behavior? (add screenshot)
clipboard02-ar

What went wrong? (add screenshot)
clipboard01-pdfjs

### LINKED ISSUE #9928: Transparent shape missing
Attach (recommended) or Link to PDF file here:
[pop-bugzilla48453.pdf](https://github.com/mozilla/pdf.js/files/2232840/pop-bugzilla48453.pdf)

From poppler issue:
https://bugs.freedesktop.org/show_bug.cgi?id=48453
"When the gstate has both an smask and a constant opacity < 1, the cairo backend ignores the constant opacity", this may or may not apply.

Configuration:
- Web browser and its version: Microsoft Edge 42.17134.1.0
- Operating system and its version: W10 home
- PDF.js version: latest

Steps to reproduce the problem:
1. start viewer in browser
2. open file

What is the expected behavior? (add screenshot)
clipboard01-ar

What went wrong? (add screenshot)
clipboard01-pdfjs

### PR DESCRIPTION
**Pull request by Rob--W**
This PR significantly improves the performance of PDFs with tiling patterns (which also fixes a bunch of blurry images), and fixes a correctness issue caused by a missing transform.

Fixes #6496
Fixes #5698
Fixes #1434
Fixes #2825 (the rendering issues have been fixed, but there are still spikes in memory usage (which is very problematic on low-memory devices), this is probably caused by https://crbug.com/448150, and possibly the same issue as https://github.com/bp74/StageXL/issues/214?. I think that we can work around this memory spike by introducing an artificial delay in our task scheduler to give the browser a chance to GC the temporary CanvasGradient patterns - this could be added in a separate PR).

Fixes #6737
Fixes #7731
Fixes #8264
Fixes #8579
Fixes #9922
Fixes #9927
Fixes #9928
Fixes #7946
Fixes #8083
Fixes #8373
Fixes #9770
Fixes #9665
Fixes #9763
Fixes https://bugzilla.mozilla.org/show_bug.cgi?id=1467424

### CODE REVIEW

**Comment by timvandermeij** [comment]
Amazing, thank you for looking into this! I can confirm that all four issues are fixed with this PR. The PDF file in the last issue is rendering much faster (almost instantly for me) now and I don't really experience the mentioned memory spike, so I think we're fine there too.

/botio test

**Comment by Rob--W** [comment]
@timvandermeij I've found the memory spike in Chrome (on Linux), using the following steps:

1. Download the PDF file from https://github.com/mozilla/pdf.js/issues/2825#issuecomment-24234561 (to make sure that rendering happens at once, without waiting for network)
2. Open the PDF.js PDF Viewer.
3. Open the task manager of Chrome (Shift + Esc) and keep an eye on the tab with PDF.js
4. In the Open PDF dialog, select the PDF file that you downloaded from step 1.
5. While the PDF is being rendered, memory usage shoots up to over 1 GB. Shortly after completion of rendering, the memory usage drops again.

**Comment by Rob--W** [comment]
The test previously failed with observable differences because I accidentally used `Math.max` instead of `Math.min`

I have updated the PR with the fix. When I ran the tests locally, 3 tests failed (chrome-High-Pressure-Measurement-WP-001287-page3,  chrome-colorkeymask-page1, chrome-issue2177-eq-page1). These failures are caused by slight differences in rounding errors (previously we manually calculated the scale, now most of the scale calculations is delegated to the canvas). I don't see any visual problems with the rendering, so you have to regenerate the refs for these three test cases after merging this PR.

File: src/display/pattern_helper.js [hunk 1]
```diff
@@ -322,63 +322,58 @@ var TilingPattern = (function TilingPatternClosure() {
 
       var x0 = bbox[0], y0 = bbox[1], x1 = bbox[2], y1 = bbox[3];
 
-      var topLeft = [x0, y0];
-      // we want the canvas to be as large as the step size
-      var botRight = [x0 + xstep, y0 + ystep];
+      // It is a waste of memory to use the size as specified by xstep or ystep
+      // if the bounding box is smaller.
+      xstep = Math.min(xstep, x1);
```
Comment (yurydelendik) [inline src/display/pattern_helper.js]: x0 and x1 are left and right coordinates of the box. shall it be `x1 - x0` here?

File: src/display/pattern_helper.js [hunk 2]
```diff
@@ -322,63 +322,58 @@ var TilingPattern = (function TilingPatternClosure() {
...
       this.setFillAndStrokeStyleToContext(tmpCtx, paintType, color);
 
-      this.setScale(width, height, xstep, ystep);
-      this.transformToScale(graphics);
-
       // transform coordinates to pattern space
-      var tmpTranslate = [1, 0, 0, 1, -topLeft[0], -topLeft[1]];
-      graphics.transform.apply(graphics, tmpTranslate);
-
+      graphics.transform(dimx.scale, 0, 0, dimy.scale, 0, 0);
```
Comment (yurydelendik) [inline src/display/pattern_helper.js]: shall it have additional -x0, -y0 translate?

File: src/display/pattern_helper.js (hunk 1 above)
Comment (Rob--W) [inline src/display/pattern_helper.js]: I don't think so (correct me if I'm wrong):
The drawing is clipped below (by `clipBbox`) by the square (x0,y0) - (x1,y1). So the canvas itself must be at least as large as (x1,y1).

Using the difference `x1 - x0` would be possible if we can translate the viewport and clip at the rectangle (0,0) - (x1-x0, y1-y0). I haven't tried this yet, do you think that this is a valid operation?

File: src/display/pattern_helper.js (hunk 2 above)
Comment (Rob--W) [inline src/display/pattern_helper.js]: No, I removed it because the transformation actually resulted in a bad rendering: fad83a7deb2d3cb1b6597f11e4e5ecfd7e9a3c94.

File: src/display/pattern_helper.js (hunk 1 above)
Comment (yurydelendik) [inline src/display/pattern_helper.js]: yeah, that's what translation -x0, -y0 below shall do... let me check why fad83a7 fixes the issue

Comment (yurydelendik) [inline src/display/pattern_helper.js]: lots of mistakes was in the initial code, let's see if we can fix them:
- we shall rely on pattern cell width/height not on its clip boundary coordinate or xstep -- the mistake was we used xstep instead of pattern width, I don't think xstep shall participate in the calculations
- once we translated by -x0,-y0, we shall do the reverse before painting (at the scaleToContext spot)

Comment (yurydelendik) [inline src/display/pattern_helper.js]: See also http://www.adobe.com/content/dam/Adobe/en/devnet/acrobat/pdfs/PDF32000_2008.pdf#page=183&zoom=180,23,420 

Also notice XStep/YStep comments below:
- "XStep  and  YStep  may  differ  from  the  dimensions  of  the  pattern  cell  implied  by  the  BBox  entry."
- "XStep and YStep may be either positive or negative but shall not be zero. "

File: src/display/pattern_helper.js [hunk 3]
```diff
@@ -322,63 +322,58 @@ var TilingPattern = (function TilingPatternClosure() {
...
-        matrixScale[1] * curMatrixScale[1]];
-
-      // MAX_PATTERN_SIZE is used to avoid OOM situation.
       // Use width and height values that are as close as possible to the end
       // result when the pattern is used. Too low value makes the pattern look
       // blurry. Too large value makes it look too crispy.
-      width = Math.min(Math.ceil(Math.abs(width * combinedScale[0])),
-        MAX_PATTERN_SIZE);
+      var ctm = this.ctx.mozCurrentTransform;
+      var dimx = this.getSizeAndScale(xstep, this.ctx.canvas.width, ctm[0]);
```
Comment (yurydelendik) [inline src/display/pattern_helper.js]: Don't rely on ctm alone, use singularValueDecompose2dScale, that's why combinedScale was introduced.

File: src/display/pattern_helper.js [hunk 4]
```diff
@@ -322,63 +322,58 @@ var TilingPattern = (function TilingPatternClosure() {
...
+      this.ctx.scale(1 / dimx.scale, 1 / dimy.scale);
+      return tmpCanvas.canvas;
     },
 
-    scaleToContext: function TilingPattern_scaleToContext() {
-      var scale = this.scale;
-      this.ctx.scale(1 / scale[0], 1 / scale[1]);
+    getSizeAndScale:
+        function TilingPattern_getSizeAndScale(step, realOutputSize, scale) {
+      // TODO(robwu): The scale should not be 0. Is this the right place to
```
Comment (yurydelendik) [inline src/display/pattern_helper.js]: See comment above about singularValueDecompose2dScale, in your case the scale parameter will be zero when `ctx.rotate(Math.PI/2)` is used.

File: src/display/pattern_helper.js (hunk 1 above)
Comment (yurydelendik) [inline src/display/pattern_helper.js]: ... in case when Math.abs(XStep) is greater than cell width, we need to grow canvas, and in case when Math.abs(XStep) less than cell width we probably need to create a pattern that is painted on itself.

The canvas 2d context does not let us to set repeat distances so we probably need to do something clever, like paint in SVG, produce a canvas and "grow" it with SVG, disable "repeat" and repeat in fill/stroke. Also we need to apply optimizations for well-know paint patterns (e.g. painting simple image as using pattern) to replace with simple primitives.

**Comment by timvandermeij** [comment]
@Rob--W Do you have time to rebase this PR and fix the comments above? It looks like a nice patch to have as it fixes quite some open issues.

**Comment by Rob--W** [comment]
@timvandermeij Last month I discussed this PR with @yurydelendik over IRC (http://logs.glob.uno/?c=mozilla%23pdfjs&s=13+Nov+2015&e=13+Nov+2015&h=6606#c41194) and the conclusion is that the implementation can be improved. This change is a significant refactor of the tiling and drawing code, and I haven't allocated time for that task yet. Of course, I can rebase and fix the nits that are mentioned above (so that we have the desired functionality) and submit a new PR for an improved version.

If @yurydelendik does not have any objections against merging this PR (with the nits fixed), then I will rebase and update the PR so you can merge it.

**Comment by Snuffleupagus** [comment]
Given that this PR completely fixes a handful of issues, and also improves a few more, could we perhaps now reconsider landing this as is (with the last round of review comments addressed)!?
It seems to me that while it's certainly possible to improve the general implementation (as outlined in the IRC discussion), this PR would still be a nice improvement of the pattern handling.

**Comment by timvandermeij** [comment]
I agree with the above. I do not know enough of this part of the codebase to review it, but if it's an improvement (perhaps not a perfect solution yet), I would also like to have this landed. It can always be improved later on.

**Comment by janpe2** [comment]
@Rob--W After addressing the review comments, the class TilingPattern could maybe look like this:
https://gist.github.com/janpe2/beb4a18d73dffb250830e7ddc41bd461
I have tested lots of PDFs and this would fix as many as 14 issues.

**Comment by timvandermeij** [comment]
Let's get this PR rolling again since it will fix quite a large number of issues. @Rob--W Do you have time to fix this up based on the comment above? If not, perhaps @janpe2 could create a new PR with the suggested changes.

**Comment by Snuffleupagus** [comment]
Given that there hasn't been any response to https://github.com/mozilla/pdf.js/pull/6606#issuecomment-455872532 for almost a month, and that the issues keep on coming, I cannot see the harm in @janpe2 (provided you want to, of course) finishing this PR so that it can land.

**Comment by Rob--W** [comment]
I'm slowly working on this.

Neither my PR nor @janpe2's PR are suitable for merging, because of one glaring issue (which I will describe below). I locally revived the patch when https://github.com/mozilla/pdf.js/pull/6606#issuecomment-455872532 was posted, but was concerned about the unaddressed issues (in https://github.com/mozilla/pdf.js/pull/6606/files/fad83a7deb2d3cb1b6597f11e4e5ecfd7e9a3c94#r44318712).

To check whether there are any potential regressions, I was looking for a test case (PDF) that has a pattern where the xstep that is smaller than the bounding box's width (or equivalently, ystep and height), AND that repeats that pattern.
In the current version of the PR, the width/height of the pattern canvas is set to the xstep/ystep (whereas it was previously set to the pattern's bounding box). The expected common case is where `xstep > width` (or `ystep > height`). This is rendered well in the current PR.
In the case where `xstep < width` (or `ystep < height`) AND the pattern is repeated, part of the pattern may be cut off.

For example, a `V` pattern with the xstep set to half of its width.
```
Base pattern:
\/

Expected rendering of repeated pattern (recall xstep = half of width):
\XXXX

Actual rendering if truncating pattern width to xstep:
\\\\\
```

To check whether the issue is real, and to confirm a fix, I need an actual test case (PDF). I'm probably going to create such a test case by hand, since searching for an existing test case with this issue is like looking for a needle in a haystack.

**Comment by THausherr** [comment]
@Rob--W 
[patterns-Xstep5.pdf](https://github.com/mozilla/pdf.js/files/2879161/patterns-Xstep5.pdf)
[patterns-Xstep10.pdf](https://github.com/mozilla/pdf.js/files/2879163/patterns-Xstep10.pdf)

**Comment by THausherr** [comment]
@Rob--W 
[patterns-Xstep5.pdf](https://github.com/mozilla/pdf.js/files/2879161/patterns-Xstep5.pdf)
[patterns-Xstep10.pdf](https://github.com/mozilla/pdf.js/files/2879163/patterns-Xstep10.pdf)
[patterns-Xstep7.pdf](https://github.com/mozilla/pdf.js/files/2879197/patterns-Xstep7.pdf)

**Comment by janpe2** [comment]
This is what the standard says about overlapping tiles (section 8.7.3.1):
> The order in which individual tiles (instances of the cell) are painted is unspecified and unpredictable; figures on adjacent tiles should not overlap.

However, Acrobat Reader draws them with no problems.

---

If someone is interested, here are the issues that would be fixed by my proposed patch:
#6496
#1434
#2825
#8264
#7731
#9763
#9927
#9928
#9922
#8579
#9665
#8083
#8373
[Bugzilla-1467424](https://bugzilla.mozilla.org/show_bug.cgi?id=1467424)

**Comment by Rob--W** [comment]
@THausherr  Thanks for sharing these test cases. That really saves me some time!
@janpe2  Thanks for checking those. I expect that the final version of this PR will fix those bugs, and avoid the regression that I described in https://github.com/mozilla/pdf.js/pull/6606#issuecomment-464784033

**Comment by THausherr** [comment]
Here's another. The shape to the top and to the right should be "pointy" and not flat when rendering is correct.
[gs-bugzilla690487.pdf](https://github.com/mozilla/pdf.js/files/2881471/gs-bugzilla690487.pdf)

The file from #9770 does also have a "too small" step.

@Rob--W your comment from a day ago also showed that there's a bug in PDFBox 😭

**Comment by Rob--W** [comment]
FYI I was and am still traveling, and will not update this PR before 16 March. Thanks for your patience.

**Comment by Rob--W** [comment]
I have updated the PR, and while addressing the review I also tried to minimize the changes compared to the current master. I did not fix the issue that I described in https://github.com/mozilla/pdf.js/pull/6606#issuecomment-464784033 because the issue already existed before my PR (so it is not a regression), and addressing it is complicated and may result in worse performance.

Also, the specific test case that I described is not a valid PDF: Tiling patterns may overlap, but the figures on the pattern cells are not allowed to overlap by the PDF spec.

The test that I added passes, as well as the ones that I referenced in the original first comment on this PR. Please test whether this fixes other issues (@ contributors on this thread) and merge it if it looks good (@timvandermeij ).

**Comment by timvandermeij** [comment]
Thank you for fixing this up! I see that Travis reports some linting issues; could you fix those? For line 412, you should be able to use the shorthand `return { scale, size, };` instead to solve those errors.

**Comment by Rob--W** [comment]
@timvandermeij Done.

**Comment by timvandermeij** [comment]
I'm just wondering, should we add a test case for commit https://github.com/mozilla/pdf.js/pull/6606/commits/c4c92da57143ab39fe1c78760206a0e67a21e876 or is that also covered by the test case you have already added in another commit here?

**Comment by Rob--W** [comment]
The test case of the third commit already covered the second patch mostly: The PDF consists of 63 patterns with bbox `[0 0 596 842]` (same as page size), and xstep = ystep = 2876. Before my second patch, the result would be blurry in Firefox and cause an out-of-memory crash in Chrome. With the patch, the rendering is sharp.

I added a new test anyway, a minimal file that tests the relevant edge cases that caused this bug. It also features another change compared to previous code: I'm using `Math.max(MAX_PATTERN_SIZE, realOutputSize)` instead of `Math.min`, because otherwise the pattern would be clipped if the output canvas size is larger than `MAX_PATTERN_SIZE` (e.g. when the user uses 400% zoom).

**Comment by Rob--W** [comment]
I updated the first comment with a list of bugs that can be closed once this is merged, from the list of issues with the `4-tiling-type` or `4-shading-pattern` label, or were listed in an earlier comment:

- https://github.com/mozilla/pdf.js/labels/4-tiling-type
- https://github.com/mozilla/pdf.js/issues?q=is%3Aopen+label%3A4-shading-pattern+-label%3A4-tiling-type
- https://github.com/mozilla/pdf.js/pull/6606#issuecomment-465224776

**Review by timvandermeij — APPROVED** [review]
Nice work fixing this many issues! The commits are clear and the extra comments are helpful. The increased reference test coverage is also really good.
